drivers/periph_common/init.c

```diff
@@ -60,6 +63,52 @@
 #ifdef MODULE_PERIPH_INIT_SDMMC
 #include "sdmmc/sdmmc.h"
 #endif
+
+#if IS_USED(MODULE_PERIPH_INIT_I2C)
+#include "periph/gpio.h"
+#include "busy_wait.h"
+
+static void _i2c_sync_devices(i2c_t dev)
+{
+    /* Recover possible bus hangup by clocking peripheral
+     * i2c device state machines into idle.
+     * Hangup can occur if a transaction was interrupted
+     * by a reset of the MCU.
+     * Badly designed hardware can also cause a hangup
+     * due to glitches on the bus during operation which
+     * we do not aim to mitigate here.
+     */
+    gpio_init(i2c_config[dev].sda_pin, GPIO_IN);
+    gpio_init(i2c_config[dev].scl_pin, GPIO_OUT);
```

Comment:
I'm afraid this will not work on all platforms as `sda_pin`/`scl_pin` is not used everywhere.

You can use `i2c_pin_sda()`/`i2c_pin_scl()` instead, but this requires the `periph_i2c_reconfigure` feature.